README.md
[![License](https://img.shields.io/badge/License-MIT-blue.svg)](https://opensource.org/licenses/MIT)
[![Build Status](https://dev.azure.com/ariyahidayat/OpenSource/_apis/build/status/ariya.FastLZ?branchName=master)](https://dev.azure.com/ariyahidayat/OpenSource/_build/latest?definitionId=22&branchName=master)

## Overview

FastLZ (MIT license) is an ANSI C/C90 implementation of [Lempel-Ziv 77 algorithm](https://en.wikipedia.org/wiki/LZ77_and_LZ78#LZ77) (LZ77) of lossless data compression. It is suitable to compress series of text/paragraphs, sequences of raw pixel data, or any other blocks of data with lots of repetition. It is not intended to be used on images, videos, and other formats of data typically already in an optimal compressed form.

The focus for FastLZ is a very fast compression and decompression, doing that at the cost of the compression ratio. As an illustration, the comparison with zlib when compressing [enwik8](http://www.mattmahoney.net/dc/textdata.html) (also in [more details](https://github.com/inikep/lzbench)):

||Ratio|Compression|Decompression
|--|--|--|--|
|FastLZ |54.2%|11.1 ns/byte|6.6 ns/byte|
|zlib -1|42.3%|39.7 ns/byte|8.9 ns/byte|
|zlib -9|36.5%|136.2 ns/byte|8.8 ns/byte|

FastLZ is used by many software products, from a number of games (such as [Death Stranding](https://en.wikipedia.org/wiki/Death_Stranding)) to various open-source projects ([Godot Engine](https://godotengine.org/), [Facebook HHVM](https://hhvm.com/), [Apache Traffic Server](https://trafficserver.apache.org/), [Calligra Office](https://www.calligra.org/), [OSv](http://osv.io/), [Netty](https://netty.io/), etc). It even serves as the basis for other compression projects like [BLOSC](https://blosc.org/).

For other implementations of byte-aligned LZ77, take a look at [LZ4](https://lz4.github.io/lz4/), [Snappy](http://google.github.io/snappy/), [Density](https://github.com/centaurean/density), [LZO](http://www.oberhumer.com/opensource/lzo/), [LZF](http://oldhome.schmorp.de/marc/liblzf.html), [LZJB](https://en.wikipedia.org/wiki/LZJB), [LZRW](http://www.ross.net/compression/lzrw1.html), etc.

## Usage

FastLZ can be used directly in any C/C++ applications. For other programming languages/environments, use the corresponding binding:

* [Rust](https://crates.io/crates/fastlz), available on Crates: `cargo install fastlz`
* [Python](https://pypi.org/project/fastlz/), available on PyPi: `pip install fastlz`
* [JavaScript](https://www.npmjs.com/package/fastlz), available on npm: `npm install fastlz`
* [Ruby](https://rubygems.org/gems/fastlz), available on Rubygems: `gem install fastlz`
* Lua via [github.com/oneoo/lua-fastlz](https://github.com/oneoo/lua-fastlz)

FastLZ consists of only two files: `fastlz.h` and `fastlz.c`. Just add these files to your project in order to use FastLZ. For the detailed information on the API to perform compression and decompression, see `fastlz.h`.

For [Vcpkg](https://github.com/microsoft/vcpkg) users, FastLZ is [already available](https://github.com/microsoft/vcpkg): `vcpkg install fastlz`.

A simple file compressor called `6pack` is included as an example on how to use FastLZ. The corresponding decompressor is `6unpack`.

FastLZ supports any standard-conforming ANSI C/C90 compiler, including the popular ones such as GCC, Clang, Visual Studio and even Tiny CC. FastLZ works well on a number of architectures, from Intel/AMD (32-bit and 64-bit), ARM (32-bit and 64-bit), and MIPS.

[![Build Status](https://dev.azure.com/ariyahidayat/OpenSource/_apis/build/status/ariya.FastLZ?branchName=master)](https://dev.azure.com/ariyahidayat/OpenSource/_build/latest?definitionId=22&branchName=master)

The continuous integration system runs an extensive set of compression-decompression round trips on the following systems:


<table>
<tbody>
<tr align=center>
  <td></td>
  <td></td>
  <td>Clang</td>
  <td>GCC</td>
  <td>TinyCC</td>
  <td>VS 2017</td>
  <td>VS 2019</td>
</tr>
<tr align=center>
  <td rowspan=3>amd64</td>
  <td>Linux</td>
  <td>&#x2714;</td>
  <td>&#x2714;</td>
  <td>&#x2714;</td>
  <td></td>
  <td></td>
</tr>
<tr align=center>
  <td>macOS</td>
  <td>&#x2714;</td>
  <td>&#x2714;</td>
  <td></td>
  <td></td>
  <td></td>
</tr>
<tr align=center>
  <td>Windows</td>
  <td>&#x2714;</td>
  <td>&#x2714;</td>
  <td>&#x2714;</td>
  <td>&#x2714;</td>
  <td>&#x2714;</td>
</tr>
<tr align=center>
  <td rowspan=3>i686</td>
  <td>Linux</td>
  <td>&#x2714;</td>
  <td>&#x2714;</td>
  <td>&#x2714;</td>
  <td></td>
  <td></td>
</tr>
<tr align=center>
  <td>macOS</td>
  <td></td>
  <td></td>
  <td></td>
  <td></td>
  <td></td>
</tr>
<tr align=center>
  <td>Windows</td>
  <td>&#x2714;</td>
  <td>&#x2714;</td>
  <td>&#x2714;</td>
  <td>&#x2714;</td>
  <td>&#x2714;</td>
<tr>
<tr align=center>
  <td>arm64</td>
  <td>Linux</td>
  <td></td>
  <td>&#x2714;</td>
  <td></td>
  <td></td>
  <td></td>
</tr>
<tr align=center>
  <td>armhf</td>
  <td>Linux</td>
  <td></td>
  <td>&#x2714;</td>
  <td></td>
  <td></td>
  <td></td>
</tr>
<tr align=center>
  <td>mips</td>
  <td>Linux</td>
  <td></td>
  <td>&#x2714;</td>
  <td></td>
  <td></td>
  <td></td>
</tr>
</tbody>
</table>

## Block Format

Let us assume that FastLZ compresses an array of bytes, called the _uncompressed block_, into another array of bytes, called the _compressed block_. To understand what will be stored in the compressed block, it is illustrative to demonstrate how FastLZ will _decompress_ the block to retrieve the original uncompressed block.

The first 5-bit of the block, i.e. the 5 most-significant bits of the first byte, is the **block tag**. Currently the block tag determines the compression level used to produce the compressed block.

|Block tag|Compression level|
|---------|-----------------|
|   0     |    Level 1      |
|   1     |    Level 2      |

The content of the block will vary depending on the compression level.

### Block Format for Level 1

FastLZ Level 1 impements LZ77 compression algorithm with 8 KB sliding window and up to 264 bytes of match length.

The compressed block consists of one or more **instructions**.
Each instruction starts with a 1-byte opcode, 2-byte opcode, or 3-byte opcode.

| Instruction type | Opcode[0] | Opcode[1] | Opcode[2]
|-----------|------------------|--------------------|--|
| Literal run | `000`, L&#x2085;-L&#x2080; | -|- |
| Short match | M&#x2082;-M&#x2080;, R&#x2081;&#x2082;-R&#x2088; | R&#x2087;-R&#x2080; | - |
| Long match | `111`, R&#x2081;&#x2082;-R&#x2088; | M&#x2087;-R&#x2080; | R&#x2087;-R&#x2080; |

Note that the _very first_ instruction in a compressed block is always a literal run.

#### Literal run instruction

For the literal run instruction, there is one or more bytes following the code. This is called the literal run.

The 5 least-significant bits of `opcode[0]`, _L_, determines the **number of literals** following the opcode. The value of 0 indicates a 1-byte literal run, 1 indicates a 2-byte literal run, and so on. The minimum literal run is 1 and the maximum literal run is 32.

The decompressor copies (_L + 1_) bytes of literal run, starting from the first one right after opcode.

_Example_: If the compressed block is a 4-byte array of `[0x02, 0x41, 0x42, 0x43]`, then the opcode is `0x02` and that means a literal run of 3 bytes. The decompressor will then copy the subsequent 3 bytes, `[0x41, 0x42, 0x43]`, to the output buffer. The output buffer now represents the (original) uncompressed block, `[0x41, 0x42, 0x43]`.

#### Short match instruction

The 3 most-significant bits of `opcode[0]`, _M_, determines the **match length**. The value of 1 indicates a 3-byte match, 2 indicates a 4-byte match and so on. The minimum match length is 3 and the maximum match length is 8.

The 5 least-significant bits of `opcode[0]` combined with the 8 bits of the `opcode[1]`, _R_, determines the **reference offset**. Since the offset is encoded in 13 bits, the minimum is 0 and the maximum is 8191.

The following C code retrieves the match length and reference offset:

```c
M = opcode[0] >> 5;
R = 256 * (opcode[0] << 5) + opcode[1];
```

The decompressor copies _(M+2)_ bytes, starting from the location offsetted by _R_ in the output buffer. Note that _R_ is a *back reference*, i.e. the value of 0 corresponds the last byte in the output buffer, 1 is the second to last byte, and so forth.

_Example 1_: If the compressed block is a 7-byte array of `[0x03, 0x41, 0x42, 0x43, 0x44, 0x20, 0x02]`, then there are two instructions in the there. The first instruction is the literal run of 4 bytes (due to _L = 3_). Thus, the decompressor copies 4 bytes to the output buffer, resulting in `[0x41, 0x42, 0x43, 0x44]`. The second instruction is the short match of 3 bytes (from _M = 1_, i.e `0x20 >> 5`) and the offset of 2. Therefore, the compressor goes back 2 bytes from the last position, copies 3 bytes (`[0x42, 0x43, 0x44]`), and appends them to the output buffer. The output buffer now represents the complete uncompressed data, `[0x41, 0x42, 0x43, 0x44, 0x42, 0x43, 0x44]`.

_Example 2_: If the compressed block is a 4-byte array of `[0x00, 0x61, 0x40, 0x00]`, then there are two instructions in there. The first instruction is the literal run of just 1 byte (_L = 0_). Thus, the decompressor copies the byte (`0x61`) to the output buffer. The output buffer now becomes `[0x61]`. The second instruction is the short match of 4 bytes (from _M = 2_, i.e. `0x40 >> 5`) and the offset of 0. Therefore, the decompressor copies 4 bytes starting using the back reference of 0 (i.e. the position of `0x61`). The output buffer now represents the complete uncompressed data, `[0x61, 0x61, 0x61, 0x61, 0x61]`.

#### Long match instruction

The value of `opcode[1]`, _M_, determines the **match length**. The value of 0 indicates a 9-byte match, 1 indicates a 10-byte match and so on. The minimum match length is 9 and the maximum match length is 264.

The 5 least-significant bits of `opcode[0]` combined with the 8 bits of `opcode[2]`, _R_, determines the **reference offset**. Since the offset is encoded in 13 bits, the minimum is 0 and the maximum is 8191.

The following C code retrieves the match length and reference offset:

```c
M = opcode[1];
R = 256 * (opcode[0] << 5) + opcode[2];
```
The decompressor copies _(M+9)_ bytes, starting from the location offsetted by _R_ in the output buffer. Note that _R_ is a *back reference*, i.e. the value of 0 corresponds to the last byte in the output buffer, 1 is for the second to last byte, and so forth.

_Example_:  If the compressed block is a 4-byte array of `[0x01, 0x44, 0x45, 0xE0, 0x01, 0x01]`, then there are two instructions in there. The first instruction is the literal run with the length of 2 (due to _L = 1_). Thus, the decompressor copies the 2-byte literal run (`[0x44, 0x45]`) to the output buffer. The second instruction is the long match with the match length of 10 (from _M = 1_) and the offset of 1. Therefore, the decompressor copies 10 bytes starting using the back reference of 1 (i.e. the position of `0x44`). The output buffer now represents the complete uncompressed data, `[0x44, 0x45, 0x44, 0x45, 0x44, 0x45, 0x44, 0x45, 0x44, 0x45, 0x44, 0x45]`.

#### Decompressor Reference Implementation

The following 40-line C function implements a fully-functional decompressor for the above block format. Note that it is intended to be educational, e.g. no bound check is implemented, and therefore it is absolutely **unsafe** for production.

```c
void fastlz_level1_decompress(const uint8_t* input, int length, uint8_t* output) {
  int src = 0;
  int dest = 0;
  while (src < length) {
    int type = input[src] >> 5;
    if (type == 0) {
      /* literal run */
      int run = 1 + input[src];
      src = src + 1;
      while (run > 0) {
        output[dest] = input[src];
        src = src + 1;
        dest = dest + 1;
        run = run - 1;
      }
    } else if (type < 7) {
      /* short match */
      int ofs = 256 * (input[src] & 31) + input[src + 1];
      int len = 2 + (input[src] >> 5);
      src = src + 2;
      int ref = dest - ofs - 1;
      while (len > 0) {
        output[dest] = output[ref];
        ref = ref + 1;
        dest = dest + 1;
        len = len - 1;
      }
    } else {
      /* long match */
      int ofs = 256 * (input[src] & 31) + input[src + 2];
      int len = 9 + input[src + 1];
      src = src + 3;
      int ref = dest - ofs - 1;
      while (len > 0) {
        output[dest] = output[ref];
        ref = ref + 1;
        dest = dest + 1;
        len = len - 1;
      }
    }
  }
}
```

### Block Format for Level 2

(To be written)


ChangeLog
2020-01-10: Version 0.4.0

    Only code & infrastructure clean-up, no new functionality.


fastlz.h
/*
  FastLZ - Byte-aligned LZ77 compression library
  Copyright (C) 2005-2020 Ariya Hidayat <ariya.hidayat@gmail.com>

  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:

  The above copyright notice and this permission notice shall be included in
  all copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  THE SOFTWARE.
*/

#ifndef FASTLZ_H
#define FASTLZ_H

#define FASTLZ_VERSION 0x000400

#define FASTLZ_VERSION_MAJOR 0
#define FASTLZ_VERSION_MINOR 4
#define FASTLZ_VERSION_REVISION 0

#define FASTLZ_VERSION_STRING "0.4.0"

#if defined(__cplusplus)
extern "C" {
#endif

/**
  Compress a block of data in the input buffer and returns the size of
  compressed block. The size of input buffer is specified by length. The
  minimum input buffer size is 16.

  The output buffer must be at least 5% larger than the input buffer
  and can not be smaller than 66 bytes.

  If the input is not compressible, the return value might be larger than
  length (input buffer size).

  The input buffer and the output buffer can not overlap.

  Compression level can be specified in parameter level. At the moment,
  only level 1 and level 2 are supported.
  Level 1 is the fastest compression and generally useful for short data.
  Level 2 is slightly slower but it gives better compression ratio.

  Note that the compressed data, regardless of the level, can always be
  decompressed using the function fastlz_decompress below.
*/

int fastlz_compress_level(int level, const void* input, int length,
                          void* output);

/**
  Decompress a block of compressed data and returns the size of the
  decompressed block. If error occurs, e.g. the compressed data is
  corrupted or the output buffer is not large enough, then 0 (zero)
  will be returned instead.

  The input buffer and the output buffer can not overlap.

  Decompression is memory safe and guaranteed not to write the output buffer
  more than what is specified in maxout.

  Note that the decompression will always work, regardless of the
  compression level specified in fastlz_compress_level above (when
  producing the compressed block).
 */

int fastlz_decompress(const void* input, int length, void* output, int maxout);

/**
  DEPRECATED.

  This is similar to fastlz_compress_level above, but with the level
  automatically chosen.

  This function is deprecated and it will be completely removed in some future
  version.
*/

int fastlz_compress(const void* input, int length, void* output);

#if defined(__cplusplus)
}
#endif

#endif /* FASTLZ_H */


examples/6pack.c
/*
  6PACK - file compressor using FastLZ (lightning-fast compression library)
  Copyright (C) 2007-2020 Ariya Hidayat <ariya.hidayat@gmail.com>

  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:

  The above copyright notice and this permission notice shall be included in
  all copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  THE SOFTWARE.
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define SIXPACK_VERSION_MAJOR 0
#define SIXPACK_VERSION_MINOR 1
#define SIXPACK_VERSION_REVISION 0
#define SIXPACK_VERSION_STRING "snapshot 20070615"

#include "fastlz.h"

#undef PATH_SEPARATOR

#if defined(MSDOS) || defined(__MSDOS__) || defined(MSDOS)
#define PATH_SEPARATOR '\\'
#endif

#if defined(WIN32) || defined(__NT__) || defined(_WIN32) || defined(__WIN32__)
#define PATH_SEPARATOR '\\'
#endif

#ifndef PATH_SEPARATOR
#define PATH_SEPARATOR '/'
#endif

#undef SIXPACK_BENCHMARK_WIN32
#if defined(WIN32) || defined(__NT__) || defined(_WIN32) || defined(__WIN32__)
#if defined(_MSC_VER) || defined(__GNUC__)
#define SIXPACK_BENCHMARK_WIN32
#include <windows.h>
#endif
#endif

/* magic identifier for 6pack file */
static unsigned char sixpack_magic[8] = {137, '6', 'P', 'K', 13, 10, 26, 10};

#define BLOCK_SIZE (2 * 64 * 1024)

/* prototypes */
static unsigned long update_adler32(unsigned long checksum, const void* buf,
                                    int len);
void usage(void);
int detect_magic(FILE* f);
void write_magic(FILE* f);
void write_chunk_header(FILE* f, int id, int options, unsigned long size,
                        unsigned long checksum, unsigned long extra);
unsigned long block_compress(const unsigned char* input, unsigned long length,
                             unsigned char* output);
int pack_file_compressed(const char* input_file, int method, int level,
                         FILE* f);
int pack_file(int compress_level, const char* input_file,
              const char* output_file);

/* for Adler-32 checksum algorithm, see RFC 1950 Section 8.2 */
#define ADLER32_BASE 65521
static unsigned long update_adler32(unsigned long checksum, const void* buf,
                                    int len) {
  const unsigned char* ptr = (const unsigned char*)buf;
  unsigned long s1 = checksum & 0xffff;
  unsigned long s2 = (checksum >> 16) & 0xffff;

  while (len > 0) {
    unsigned k = len < 5552 ? len : 5552;
    len -= k;

    while (k >= 8) {
      s1 += *ptr++;
      s2 += s1;
      s1 += *ptr++;
      s2 += s1;
      s1 += *ptr++;
      s2 += s1;
      s1 += *ptr++;
      s2 += s1;
      s1 += *ptr++;
      s2 += s1;
      s1 += *ptr++;
      s2 += s1;
      s1 += *ptr++;
      s2 += s1;
      s1 += *ptr++;
      s2 += s1;
      k -= 8;
    }

    while (k-- > 0) {
      s1 += *ptr++;
      s2 += s1;
    }
    s1 = s1 % ADLER32_BASE;
    s2 = s2 % ADLER32_BASE;
  }
  return (s2 << 16) + s1;
}

void usage(void) {
  printf("6pack: high-speed file compression tool\n");
  printf("Copyright (C) Ariya Hidayat\n");
  printf("\n");
  printf("Usage: 6pack [options]  input-file  output-file\n");
  printf("\n");
  printf("Options:\n");
  printf("  -1    compress faster\n");
  printf("  -2    compress better\n");
  printf("  -v    show program version\n");
#ifdef SIXPACK_BENCHMARK_WIN32
  printf("  -mem  check in-memory compression speed\n");
#endif
  printf("\n");
}

/* return non-zero if magic sequence is detected */
/* warning: reset the read pointer to the beginning of the file */
int detect_magic(FILE* f) {
  unsigned char buffer[8];
  size_t bytes_read;
  int c;

  fseek(f, SEEK_SET, 0);
  bytes_read = fread(buffer, 1, 8, f);
  fseek(f, SEEK_SET, 0);
  if (bytes_read < 8) return 0;

  for (c = 0; c < 8; c++)
    if (buffer[c] != sixpack_magic[c]) return 0;

  return -1;
}

void write_magic(FILE* f) { fwrite(sixpack_magic, 8, 1, f); }

void write_chunk_header(FILE* f, int id, int options, unsigned long size,
                        unsigned long checksum, unsigned long extra) {
  unsigned char buffer[16];

  buffer[0] = id & 255;
  buffer[1] = id >> 8;
  buffer[2] = options & 255;
  buffer[3] = options >> 8;
  buffer[4] = size & 255;
  buffer[5] = (size >> 8) & 255;
  buffer[6] = (size >> 16) & 255;
  buffer[7] = (size >> 24) & 255;
  buffer[8] = checksum & 255;
  buffer[9] = (checksum >> 8) & 255;
  buffer[10] = (checksum >> 16) & 255;
  buffer[11] = (checksum >> 24) & 255;
  buffer[12] = extra & 255;
  buffer[13] = (extra >> 8) & 255;
  buffer[14] = (extra >> 16) & 255;
  buffer[15] = (extra >> 24) & 255;

  fwrite(buffer, 16, 1, f);
}

int pack_file_compressed(const char* input_file, int method, int level,
                         FILE* f) {
  FILE* in;
  unsigned long fsize;
  unsigned long checksum;
  const char* shown_name;
  unsigned char buffer[BLOCK_SIZE];
  unsigned char result[BLOCK_SIZE * 2]; /* FIXME twice is too large */
  unsigned char progress[20];
  int c;
  unsigned long percent;
  unsigned long total_read;
  unsigned long total_compressed;
  int chunk_size;

  /* sanity check */
  in = fopen(input_file, "rb");
  if (!in) {
    printf("Error: could not open %s\n", input_file);
    return -1;
  }

  /* find size of the file */
  fseek(in, 0, SEEK_END);
  fsize = ftell(in);
  fseek(in, 0, SEEK_SET);

  /* already a 6pack archive? */
  if (detect_magic(in)) {
    printf("Error: file %s is already a 6pack archive!\n", input_file);
    fclose(in);
    return -1;
  }

  /* truncate directory prefix, e.g. "foo/bar/FILE.txt" becomes "FILE.txt" */
  shown_name = input_file + strlen(input_file) - 1;
  while (shown_name > input_file)
    if (*(shown_name - 1) == PATH_SEPARATOR)
      break;
    else
      shown_name--;

  /* chunk for File Entry */
  buffer[0] = fsize & 255;
  buffer[1] = (fsize >> 8) & 255;
  buffer[2] = (fsize >> 16) & 255;
  buffer[3] = (fsize >> 24) & 255;
#if 0
  buffer[4] = (fsize >> 32) & 255;
  buffer[5] = (fsize >> 40) & 255;
  buffer[6] = (fsize >> 48) & 255;
  buffer[7] = (fsize >> 56) & 255;
#else
  /* because fsize is only 32-bit */
  buffer[4] = 0;
  buffer[5] = 0;
  buffer[6] = 0;
  buffer[7] = 0;
#endif
  buffer[8] = (strlen(shown_name) + 1) & 255;
  buffer[9] = (strlen(shown_name) + 1) >> 8;
  checksum = 1L;
  checksum = update_adler32(checksum, buffer, 10);
  checksum = update_adler32(checksum, shown_name, strlen(shown_name) + 1);
  write_chunk_header(f, 1, 0, 10 + strlen(shown_name) + 1, checksum, 0);
  fwrite(buffer, 10, 1, f);
  fwrite(shown_name, strlen(shown_name) + 1, 1, f);
  total_compressed = 16 + 10 + strlen(shown_name) + 1;

  /* for progress status */
  memset(progress, ' ', 20);
  if (strlen(shown_name) < 16)
    for (c = 0; c < (int)strlen(shown_name); c++) progress[c] = shown_name[c];
  else {
    for (c = 0; c < 13; c++) progress[c] = shown_name[c];
    progress[13] = '.';
    progress[14] = '.';
    progress[15] = ' ';
  }
  progress[16] = '[';
  progress[17] = 0;
  printf("%s", progress);
  for (c = 0; c < 50; c++) printf(".");
  printf("]\r");
  printf("%s", progress);

  /* read file and place in archive */
  total_read = 0;
  percent = 0;
  for (;;) {
    int compress_method = method;
    int last_percent = (int)percent;
    size_t bytes_read = fread(buffer, 1, BLOCK_SIZE, in);
    if (bytes_read == 0) break;
    total_read += bytes_read;

    /* for progress */
    if (fsize < (1 << 24))
      percent = total_read * 100 / fsize;
    else
      percent = total_read / 256 * 100 / (fsize >> 8);
    percent >>= 1;
    while (last_percent < (int)percent) {
      printf("#");
      last_percent++;
    }

    /* too small, don't bother to compress */
    if (bytes_read < 32) compress_method = 0;

    /* write to output */
    switch (compress_method) {
      /* FastLZ */
      case 1:
        chunk_size = fastlz_compress_level(level, buffer, bytes_read, result);
        checksum = update_adler32(1L, result, chunk_size);
        write_chunk_header(f, 17, 1, chunk_size, checksum, bytes_read);
        fwrite(result, 1, chunk_size, f);
        total_compressed += 16;
        total_compressed += chunk_size;
        break;

      /* uncompressed, also fallback method */
      case 0:
      default:
        checksum = 1L;
        checksum = update_adler32(checksum, buffer, bytes_read);
        write_chunk_header(f, 17, 0, bytes_read, checksum, bytes_read);
        fwrite(buffer, 1, bytes_read, f);
        total_compressed += 16;
        total_compressed += bytes_read;
        break;
    }
  }

  fclose(in);
  if (total_read != fsize) {
    printf("\n");
    printf("Error: reading %s failed!\n", input_file);
    return -1;
  } else {
    printf("] ");
    if (total_compressed < fsize) {
      if (fsize < (1 << 20))
        percent = total_compressed * 1000 / fsize;
      else
        percent = total_compressed / 256 * 1000 / (fsize >> 8);
      percent = 1000 - percent;
      printf("%2d.%d%% saved", (int)percent / 10, (int)percent % 10);
    }
    printf("\n");
  }

  return 0;
}

int pack_file(int compress_level, const char* input_file,
              const char* output_file) {
  FILE* f;
  int result;

  f = fopen(output_file, "rb");
  if (f) {
    fclose(f);
    printf("Error: file %s already exists. Aborted.\n\n", output_file);
    return -1;
  }

  f = fopen(output_file, "wb");
  if (!f) {
    printf("Error: could not create %s. Aborted.\n\n", output_file);
    return -1;
  }

  write_magic(f);

  result = pack_file_compressed(input_file, 1, compress_level, f);
  fclose(f);

  return result;
}

#ifdef SIXPACK_BENCHMARK_WIN32
int benchmark_speed(int compress_level, const char* input_file);

int benchmark_speed(int compress_level, const char* input_file) {
  FILE* in;
  unsigned long fsize;
  unsigned long maxout;
  const char* shown_name;
  unsigned char* buffer;
  unsigned char* result;
  size_t bytes_read;

  /* sanity check */
  in = fopen(input_file, "rb");
  if (!in) {
    printf("Error: could not open %s\n", input_file);
    return -1;
  }

  /* find size of the file */
  fseek(in, 0, SEEK_END);
  fsize = ftell(in);
  fseek(in, 0, SEEK_SET);

  /* already a 6pack archive? */
  if (detect_magic(in)) {
    printf("Error: no benchmark for 6pack archive!\n");
    fclose(in);
    return -1;
  }

  /* truncate directory prefix, e.g. "foo/bar/FILE.txt" becomes "FILE.txt" */
  shown_name = input_file + strlen(input_file) - 1;
  while (shown_name > input_file)
    if (*(shown_name - 1) == PATH_SEPARATOR)
      break;
    else
      shown_name--;

  maxout = 1.05 * fsize;
  maxout = (maxout < 66) ? 66 : maxout;
  buffer = (unsigned char*)malloc(fsize);
  result = (unsigned char*)malloc(maxout);
  if (!buffer || !result) {
    printf("Error: not enough memory!\n");
    free(buffer);
    free(result);
    fclose(in);
    return -1;
  }

  printf("Reading source file....\n");
  bytes_read = fread(buffer, 1, fsize, in);
  if (bytes_read != fsize) {
    printf("Error reading file %s!\n", shown_name);
    printf("Read %d bytes, expecting %d bytes\n", bytes_read, fsize);
    free(buffer);
    free(result);
    fclose(in);
    return -1;
  }

  /* shamelessly copied from QuickLZ 1.20 test program */
  {
    unsigned int j, y;
    size_t i, u = 0;
    double mbs, fastest;
    unsigned long compressed_size;

    printf("Setting HIGH_PRIORITY_CLASS...\n");
    SetPriorityClass(GetCurrentProcess(), HIGH_PRIORITY_CLASS);

    printf("Benchmarking FastLZ Level %d, please wait...\n", compress_level);

    i = bytes_read;
    fastest = 0.0;
    for (j = 0; j < 3; j++) {
      y = 0;
      mbs = GetTickCount();
      while (GetTickCount() == mbs)
        ;
      mbs = GetTickCount();
      while (GetTickCount() - mbs < 3000) /* 1% accuracy with 18.2 timer */
      {
        u = fastlz_compress_level(compress_level, buffer, bytes_read, result);
        y++;
      }

      mbs = ((double)i * (double)y) / ((double)(GetTickCount() - mbs) / 1000.) /
            1000000.;
      /*printf(" %.1f Mbyte/s  ", mbs);*/
      if (fastest < mbs) fastest = mbs;
    }

    printf("\nCompressed %d bytes into %d bytes (%.1f%%) at %.1f Mbyte/s.\n",
           (unsigned int)i, (unsigned int)u, (double)u / (double)i * 100.,
           fastest);

#if 1
    fastest = 0.0;
    compressed_size = u;
    for (j = 0; j < 3; j++) {
      y = 0;
      mbs = GetTickCount();
      while (GetTickCount() == mbs)
        ;
      mbs = GetTickCount();
      while (GetTickCount() - mbs < 3000) /* 1% accuracy with 18.2 timer */
      {
        u = fastlz_decompress(result, compressed_size, buffer, bytes_read);
        y++;
      }

      mbs = ((double)i * (double)y) / ((double)(GetTickCount() - mbs) / 1000.) /
            1000000.;
      /*printf(" %.1f Mbyte/s  ", mbs);*/
      if (fastest < mbs) fastest = mbs;
    }

    printf("\nDecompressed at %.1f Mbyte/s.\n\n(1 MB = 1000000 byte)\n",
           fastest);
#endif
  }

  fclose(in);
  return 0;
}
#endif /* SIXPACK_BENCHMARK_WIN32 */

int main(int argc, char** argv) {
  int i;
  int compress_level;
  int benchmark;
  char* input_file;
  char* output_file;

  /* show help with no argument at all*/
  if (argc == 1) {
    usage();
    return 0;
  }

  /* default compression level, not the fastest */
  compress_level = 2;

  /* do benchmark only when explicitly specified */
  benchmark = 0;

  /* no file is specified */
  input_file = 0;
  output_file = 0;

  for (i = 1; i <= argc; i++) {
    char* argument = argv[i];

    if (!argument) continue;

    /* display help on usage */
    if (!strcmp(argument, "-h") || !strcmp(argument, "--help")) {
      usage();
      return 0;
    }

    /* check for version information */
    if (!strcmp(argument, "-v") || !strcmp(argument, "--version")) {
      printf("6pack: high-speed file compression tool\n");
      printf("Version %s (using FastLZ %s)\n", SIXPACK_VERSION_STRING,
             FASTLZ_VERSION_STRING);
      printf("Copyright (C) Ariya Hidayat\n");
      printf("\n");
      return 0;
    }

    /* test compression speed? */
    if (!strcmp(argument, "-mem")) {
      benchmark = 1;
      continue;
    }

    /* compression level */
    if (!strcmp(argument, "-1") || !strcmp(argument, "--fastest")) {
      compress_level = 1;
      continue;
    }
    if (!strcmp(argument, "-2")) {
      compress_level = 2;
      continue;
    }

    /* unknown option */
    if (argument[0] == '-') {
      printf("Error: unknown option %s\n\n", argument);
      printf("To get help on usage:\n");
      printf("  6pack --help\n\n");
      return -1;
    }

    /* first specified file is input */
    if (!input_file) {
      input_file = argument;
      continue;
    }

    /* next specified file is output */
    if (!output_file) {
      output_file = argument;
      continue;
    }

    /* files are already specified */
    printf("Error: unknown option %s\n\n", argument);
    printf("To get help on usage:\n");
    printf("  6pack --help\n\n");
    return -1;
  }

  if (!input_file) {
    printf("Error: input file is not specified.\n\n");
    printf("To get help on usage:\n");
    printf("  6pack --help\n\n");
    return -1;
  }

  if (!output_file && !benchmark) {
    printf("Error: output file is not specified.\n\n");
    printf("To get help on usage:\n");
    printf("  6pack --help\n\n");
    return -1;
  }

#ifdef SIXPACK_BENCHMARK_WIN32
  if (benchmark)
    return benchmark_speed(compress_level, input_file);
  else
#endif
    return pack_file(compress_level, input_file, output_file);

  /* unreachable */
  return 0;
}
